### DISCUSSION TITLE
Iterate over all stores - is `_s` an API guarantee?

### DISCUSSION THREAD

**Discussion by davidAtInleague** [vuejs/pinia#1859] [Help and Questions] [answered]
Hi, I'm looking to {get a listing of, iterate over} all stores in my application. This is in support of running hooks against any stores implementing particular interfaces at certain times in the application lifecycle (onLogout, etc.). In https://github.com/vuejs/pinia/discussions/911 there seems to be a common solution of the form:

```
import { getActivePinia } from "pinia"

// map through that list and use the **$reset** fn to reset the state
getActivePinia()._s.forEach(store => store.$reset());
```
Which is a fine solution, assuming `_s` never changes its implementation's name or shape.

Is the `_s` property on Pinia instances an API guarantee? If not, can it be made so (or something similar, to enable iterating over all registered stores)?

**Comment by posva** [comment, marked as answer]
`_s` is internal so it could change anytime and it wouldn't be a major version. It's unlikely to disappear though as it's crucial internally. You can create a similar solution with a plugin:

```ts
const stores = new Set()
pinia.use(({ store }) => {
  stores.add(store)
})
```

**Comment by tintin10q** [comment]
Here is  typed solution:

```ts
import { Store } from "pinia";

const AllPiniaStores = new Set<Store>();
pinia.use(({ store }) => {
  AllPiniaStores.add(store);
  store.$all = AllPiniaStores;
  store.$reset_all = () => store.$all.forEach((s) => s.$reset());
});

declare module "pinia" {
  export interface PiniaCustomProperties {
    $all: typeof AllPiniaStores;
    /** Reset all the stores */
    $reset_all: () => void;
  }
}
```

**Reply by hjisaac** [reply]
Good :+1: , **but** this way, you give each store the ability TO RESET OTHER STORES OF THE APP. It doesn't look like a good pattern though.

**Reply by tintin10q** [reply]
I guess but that was that ability was already there with _s

**Reply by hjisaac** [reply]
No, because the "_s" is a property of the active pinia instance and not of a specific store :+1: .
